text/0000-dockerfiles.md

```diff
@@ -0,0 +1,228 @@
+- `build.toml`, with the same contents as a normal buildpack's `build.toml`, but
+  - With an additional `args` table array with `name` and `value` fields that are provided as build args to `build.Dockerfile` or `Dockerfile`
+- `launch.toml`, with the same contents as a normal buildpack's `launch.toml`, but
+  - Without the `processes` table array
+  - Without the `slices` table array
+  - With an additional `args` table array with `name` and `value` fields that are provided as build args to `run.Dockerfile` or `Dockerfile`
+
+- Either `Dockerfile` or either or both of `build.Dockerfile` and `run.Dockerfile`
+
+Support for other instruction formats, e.g., LLB JSON files, could be added in the future.
+
+`build.Dockerfile`, `run.Dockerfile`, and `Dockerfile` target the builder image, runtime base image, or both base images, respectively.
+
+If no Dockerfiles are present, `/bin/generate` may still consume build plan entries and add metadata to `build.toml`/`launch.toml`.
+
+Dockerfiles are applied to their corresponding base images after all extensions are executed and before any regular buildpacks are executed.
+Dockerfiles are applied in the order determined during buildpack detection. When multiple Dockerfiles are applied, the intermediate image generated from the application of the current Dockerfile will be provided as the `base_image` ARG to the next Dockerfile. Dockerfiles that target the run image (only) may ignore the provided `base_image` (e.g., `FROM some-other-image`). Dockerfiles that change the runtime base image may still use `COPY --from=${base_image}`.
+
+All Dockerfiles are provided with `base_image` and `build_id` args.
+The `base_image` arg allows the Dockerfile to reference the original base image.
+The `build_id` arg allows the Dockerfile to invalidate the cache after a certain layer and must be defaulted to `0`. The executor of the Dockerfile will provide the `build_id` as a UUID (this eliminates the need to track this variable). 
+When the `$build_id` arg is referenced in a `RUN` instruction, all subsequent layers will be rebuilt on the next build (as the value will change).
+
+Build args specified in `build.toml` are provided to `build.Dockerfile` or `Dockerfile` (when applied to the build-time base image).
+Build args specified in `launch.toml` are provided to `run.Dockerfile` or `Dockerfile` (when applied to the runtime base image).
+
+A runtime base image may indicate that it preserves ABI compatibility by adding the label `io.buildpacks.rebasable=true`. In the case of builder-specified Dockerfiles, `io.buildpacks.rebasable=false` is set automatically on the base image before a runtime Dockerfile is applied and must be explicitly set to `true` if desired. If multiple Dockerfiles are applied, all must set `io.buildpacks.rebasable=true` for the final value to be `true`. Rebasing an app without this label set to `true` requires passing a new `--force` flag to `pack rebase`.
```

Comment:
If I am understanding correctly a rebasing an extended app image would treat extension layers like app layers rather than like run-image layers.

How does the run image top-layer get set correctly in [`io.buildpacks.lifecycle.metadata`](https://github.com/buildpacks/spec/blob/main/platform.md#iobuildpackslifecyclemetadata-json) so that we can rebase the correct layers? If we simply passed the extended run image to export the extension layers would lost, given how this works currently.

If an extension entire switches the run image (rather than just extending the existing one) how does the stack metadata get set correctly on the app image so that new versions can be found?

Reply:
Discussed in implementation sync:

This isn't relevant for "phase 1", but in the future, when we allow the run image to be extended:
* The `analyzer` when resolving creds for the run image can record the top layer in analyzed.toml
* The `exporter` can check that the top layer in analyzed.toml is a layer in the run image it was provided
* If the top layer is found (the run image was extended, not switched), then
  * `io.buildpacks.lifecycle.metadata` `runImage.topLayer` and `runImage.reference` can be derived from analyzed.toml
* If the top layer is not found (the run image was switched, and possibly also extended), then
  * The `extender` when applying Dockerfiles needs to keep track of the top layer for each image, so that it can know when a base image has been switched; the extender will provide the top layer of the run image (prior to any extensions) in some sort of output file OR it can add a label to the extended run image e.g., `io.buildpacks.rebase.metadata`
  * `io.buildpacks.lifecycle.metadata` `runImage.topLayer` and `runImage.reference` can be derived from the output of the extender

Having typed all of that out, maybe we need to be looking at ChainID (vs DiffID) in determining whether an image was switched or extended. But the principle is the same.

cc @ekcasey

Reply:
Just to note - all of this is moot if `io.buildpacks.rebasable` is false. Then the top layer will just be the top layer in the run image that was provided to export.

Reply:
Updated in 3a26570